src/sync/remote_mongo_collection.cpp

```diff
@@ -21,461 +21,411 @@
+        auto matched_count = static_cast<int32_t>(document["matchedCount"]);
+        auto modified_count = static_cast<int32_t>(document["modifiedCount"]);
+
+        Optional<ObjectId> upserted_id;
+        auto it = document.find("upsertedId");
+        if (it != document.end()) {
+            upserted_id = static_cast<ObjectId>(document["upsertedId"]);
+        }
+        
+        return completion_block(RemoteMongoCollection::RemoteUpdateResult {
+            matched_count,
+            modified_count,
+            upserted_id
+        }, error);
+    } catch (const std::exception& e) {
+        return completion_block({}, AppError(make_error_code(JSONErrorCode::bad_bson_parse), e.what()));
+    }
 }
 
-void RemoteMongoCollection::find(const std::string& filter_json,
+void RemoteMongoCollection::find(const bson::BsonDocument& filter_bson,
                                  RemoteFindOptions options,
-                                 std::function<void(Optional<std::string>, util::Optional<AppError>)> completion_block)
+                                 std::function<void(util::Optional<bson::BsonArray>, util::Optional<AppError>)> completion_block)
 {
     try {
-        auto base_args = m_base_operation_args;
-        base_args.push_back({"query", nlohmann::json::parse(filter_json)});
-        
+        auto base_args = bson::BsonDocument(m_base_operation_args);
+        base_args["query"] = filter_bson;
+
         if (options.limit) {
-            base_args.push_back({"limit", options.limit.value()});
+            base_args["limit"] = static_cast<int32_t>(*options.limit);
         }
-        
-        if (options.projection_json) {
-            base_args.push_back({"project", nlohmann::json::parse(options.projection_json.value())});
+
+        if (options.projection_bson) {
+            base_args["project"] = *options.projection_bson;
         }
-        
-        if (options.sort_json) {
-            base_args.push_back({"sort", nlohmann::json::parse(options.sort_json.value())});
+
+        if (options.sort_bson) {
+            base_args["sort"] = *options.sort_bson;
         }
         
-        auto args = nlohmann::json({{"arguments", nlohmann::json::array({base_args})}});
+        m_service->call_function("find",
+                                 bson::BsonArray({base_args}),
+                                 m_service_name,
+                                 [completion_block](util::Optional<AppError> error, util::Optional<bson::Bson> value) {
+            if (error) {
+                return completion_block(util::none, error);
+            }
 
-        m_service.call_function("find",
-                                args.dump(),
-                                [completion_block](util::Optional<AppError> error, util::Optional<std::string> value) {
-            handle_response(error, value, completion_block);
+            return completion_block(util::some<bson::BsonArray>(static_cast<bson::BsonArray>(*value)), util::none);
```

Comment:
Are both the `util::some<bson::BsonArray>()` and `static_cast<bson::BsonArray>()` actually needed here? I would expect it to only need one or the other.